src/Sets/Line.jl

```diff
@@ -0,0 +1,455 @@
+
+Check whether a given point is contained in a line.
+
+### Input
+
+- `x` -- point/vector
+- `L` -- line
+
+### Output
+
+`true` iff `x ∈ L`.
+
+### Algorithm
+
+The point ``x`` belongs to the line ``L : p + λ⋅n`` if and only if
+``x - p`` is proportional to the direction ``n``.
+"""
+function ∈(x::AbstractVector, L::Line)
+    @assert length(x) == dim(L) "expected the point and the line to have the same dimension, " *
+                                "but they are $(length(x)) and $(dim(L)) respectively"
+    _isapprox(x, L.p) && return true
+    return first(samedir(x - L.p, L.n))
+end
+
+# FIXME temp required to fix ambiguity errors
+function ∈(x::AbstractVector{N}, L::Line{N, VN}) where {N<:Real, VN<:AbstractVector{N}}
+    @assert length(x) == dim(L) "expected the point and the line to have the same dimension, " *
+                                "but they are $(length(x)) and $(dim(L)) respectively"
+    _isapprox(x, L.p) && return true
+    return first(samedir(x - L.p, L.n))
+end
+
+"""
+    rand(::Type{Line}; [N]::Type{<:Real}=Float64, [dim]::Int=2,
+         [rng]::AbstractRNG=GLOBAL_RNG, [seed]::Union{Int, Nothing}=nothing)
+
+Create a random line.
+
+### Input
+
+- `Line` -- type for dispatch
+- `N`    -- (optional, default: `Float64`) numeric type
+- `dim`  -- (optional, default: 2) dimension
+- `rng`  -- (optional, default: `GLOBAL_RNG`) random number generator
+- `seed` -- (optional, default: `nothing`) seed for reseeding
+
+### Output
+
+A random line.
+
+### Algorithm
+
+All numbers are normally distributed with mean 0 and standard deviation 1.
+"""
+function rand(::Type{Line};
+              N::Type{<:Real}=Float64,
+              dim::Int=2,
+              rng::AbstractRNG=GLOBAL_RNG,
+              seed::Union{Int, Nothing}=nothing)
+
+    rng = reseed(rng, seed)
+    n = randn(rng, N, dim)
+    while iszero(n)
+        n = randn(rng, N, dim)
+    end
+    p = randn(rng, N, dim)
+    return Line(p, n)
+end
+
+"""
+    isempty(L::Line)
+
+Return if a line is empty or not.
+
+### Input
+
+- `L` -- line
+
+### Output
+
+`false`.
+"""
+isempty(::Line) = false
+
+"""
+    constrained_dimensions(L::Line)
+
+Return the indices in which a line is constrained.
+
+### Input
+
+- `L` -- line
+
+### Output
+
+A vector of ascending indices `i` such that the line is constrained in dimension
+`i`.
+
+### Algorithm
+
+For each coordinate ``i``, vector of the form ``x_i = p_i + λ n_i`` are constrained
+i.e. (they belong to the line and are bounded) if and only if ``n_i`` is zero.
+Hence, this function returns all indices of the normal vector ``n`` for which
+the ``i``-th coordinate is nonzero.
+
+### Examples
+
+The line ``y = 5`` in two dimensions can be written as ``p = [0, 5]`` and
+``n = [1, 0)]``. This line constrains dimension ``2``.
+
+```jldoctest
+julia> constrained_dimensions(Line([0, 5.], [1, 0.]))
+1-element Array{Int64,1}:
+ 2
+```
+"""
+function constrained_dimensions(L::Line)
+    return findall(isapproxzero, L.n)
+end
```

Comment:
I am unsure about the semantics of `constrained_dimensions` (which is not a good sign :sweat_smile:). My current understanding is that the result is those dimensions where the set is not the universe, in the sense that the set is equivalent to the Cartesian product of a projection to those dimensions and suitable low-dimensional universes. In that sense the line is always constrained, with the exception of the 1D case.

Reply:
> In that sense the line is always constrained, with the exception of the 1D case.

consider the line L in 3D which passes through the origin and the point (1, 0, 0). This line is constrained along both y and z, so i would expect constrained_dimensions to return `[2, 3]` which is the current behavior. In general, since any point of the line is of the form `x = p + a * n` where `p` is a point on the line `n` is the direction of the line and `a` is a real parameter, the line is constrained along the i-th coordinate iff `n[i]` zero.

Reply:
is this resolved?

Reply:
No. Let's discuss on Gitter.

Reply:
outsourced to https://github.com/JuliaReach/LazySets.jl/issues/2245